File: leetcode/solutions/0235.LowestCommonAncestorOfABinarySearchTree/lowestCommonAncestor.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";

export default function lowestCommonAncestor(
  root: BinaryTreeNode<number> | null,
  p: BinaryTreeNode<number> | null,
  q: BinaryTreeNode<number> | null
): BinaryTreeNode<number> | null {
  if (root === null || p === null || q === null) {
    return null;
  }

  if (root.val > p.val && root.val > q.val) {
    return lowestCommonAncestor(root.left, p, q);
  } else if (root.val < p.val && root.val < q.val) {
    return lowestCommonAncestor(root.right, p, q);
  }

  return root;
}

File: leetcode/solutions/0235.LowestCommonAncestorOfABinarySearchTree/lowestCommonAncestor.go
package solution

import (
	"../../../datastructure"
)

// TreeNode structure for LeetCode
type TreeNode = datastructure.BinaryTreeNode

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if root == p || root == q {
		return root
	}

	if root.Val > p.Val && root.Val > q.Val {
		return lowestCommonAncestor(root.Left, p, q)
	} else if root.Val < p.Val && root.Val < q.Val {
		return lowestCommonAncestor(root.Right, p, q)
	}

	return root
}
